connect/api/src/main/java/org/apache/kafka/connect/sink/SinkRecord.java

```diff
@@ -79,17 +195,21 @@ public boolean equals(Object o) {
 
         SinkRecord that = (SinkRecord) o;
 
-        if (kafkaOffset != that.kafkaOffset)
-            return false;
-
-        return timestampType == that.timestampType;
+        return kafkaOffset == that.kafkaOffset &&
+                timestampType == that.timestampType &&
+                Objects.equals(originalTopic, that.originalTopic) &&
+                Objects.equals(originalKafkaPartition, that.originalKafkaPartition)
+                && originalKafkaOffset == that.originalKafkaOffset;
```

Comment:
This change is binary compatible. Also, it shouldn't affect any existing sink connectors in any way (i.e. it is effectively source / API compatible) since the existing public constructors for `SinkRecord` have been modified to simply use `topic`, `kafkaPartition`, and `kafkaOffset` for `originalTopic`, `originalKafkaPartition`, and `originalKafkaOffset` respectively (so records that were equal earlier will continue being equal and vice versa).

Reply:
The connector can receive two SinkRecords which may pass equality if they have the same T/P/O, and the same contents, header, etc. Right now this can happen when an SMT causes a T/P/O collision, and if a record is re-delivered to the connector. After this change, only a re-delivery can cause this equality check to fire.

I don't think it would be reasonable for someone to rely on this to detect T/P/O collisions, because it is inherently unreliable due to checking equality of the contents of the record. I think the semantics for only being equal after a re-delivery is much better.

We don't seem to rely on this equals method in the runtime, so this should only affect connector implementations.